lessons/intro/regex/index.md

```diff
@@ -0,0 +1,268 @@
+<re.Match object; span=(15, 17), match='ny'>
+>>> re.findall(r".y", "Naše staré hodiny bijí čtyři hodiny.")
+['ny', 'ty', 'ny']
+>>> list(re.finditer(r".y", "Naše staré hodiny bijí čtyři hodiny."))
+[<re.Match object; span=(15, 17), match='ny'>, <re.Match object; span=(24, 26), match='ty'>, <re.Match object; span=(33, 35), match='ny'>]
+```
+
+### Nahrazování
+
+Nahrazení funkcí `sub` funguje velmi podobně jako řetězcová metoda `replace`. Samozřejmě s tím rozdílem, že nahrazená část musí
+odpovídat regulárnímu výrazu.
+
+```pycon
+>>> re.sub(r"\s+and\s+", " & ", "Bread  and butter and    ham  and  spam")
+'Bread & butter & ham & spam'
+>>> re.sub(r"\s+and\s+", " & ", "Bread  and butter and    ham  and  spam", count=1)
+'Bread & butter and    ham  and  spam'
+```
+
+V příkladu výše nahrazujeme slovo `and` obklopené jedním nebo více bílými znaky z obou stran znakem `&` s jednou mezerou na každé straně. Parametrem `count` si můžeme nastavit, kolikrát se má nahrazení provést.
+
+## Nastavení chování (flags)
+
+Jedním ze základních způsobů, jak ovlivnit chování funkcí v modulu `re` jsou tzv. flagy. Jedná se o různá nastavení,
+která je možné v případě potřeby zapnout.
+
+* `re.DEBUG` zobrazí ladící hlášky týkající se zpracování regulárního výrazu
+* `re.IGNORECASE` (zkráceně `re.I`) způsobí, že regulární výraz nebude brát ohled na velikost písmen
+* `re.MULTILINE` (zkráceně `re.M`) umožní zpracování víceřádkových řetězců. `^` a `$` začnou označovat začátek a konec každého řádku místo začátku a konce celého řetězce.
+* `re.DOTALL` (zkráceně `re.S`) způsobí, že se tečka stane zástupným znakem i za nový řádek `\n`, což v základním nastavení není.
+* … a další k nalezení v dokumentaci.
+
+Nejlepší je předávat flagy jako pojmenovaný argument výše zmíněným funkcím. Vyhneme se tím chybám vzniklým kvůli různým počtům pozičních argumentů funkcí v modulu `re`.
+Pokud jich potřebujeme předat více najednou, nakombinujeme je pomocí operátoru `|`.
+
+```pycon
+>>> re.findall(r".y", "Naše staré hodiny bijí čtyři HODINY.")
+['ny', 'ty']
+>>> re.findall(r".y", "Naše staré hodiny bijí čtyři HODINY.", flags=re.IGNORECASE)
+['ny', 'ty', 'NY']
+>>> re.findall(r".y", "Naše staré hodiny bijí čtyři HODINY.", flags=re.IGNORECASE|re.DEBUG)
+ANY None
+LITERAL 121
+
+0. INFO 4 0b0 2 2 (to 5)
+5: ANY
+6. LITERAL_UNI_IGNORE 0x79 ('y')
+8. SUCCESS
+['ny', 'ty', 'NY']
+```
+
+## Optimalizace
+
+Jednou z možností optimalizace fungování regulárních výrazů je jejich kompilace. Nejdříve si regulární výraz zkompilujete a pak místo
+funkcí z modulu `re` použijete stejnojmenné metody tohoto nového objektu.
+
+```pycon
+>>> regex = re.compile(r".y", flags=re.IGNORECASE)
+>>> regex.findall("Naše staré hodiny bijí čtyři HODINY.")
+['ny', 'ty', 'NY']
+```
+
+Kompilace se ale děje i automaticky na pozadí při používání funkcí, takže pokud používáte jen několik stále se opakujících
+regulárních výrazů, není třeba si je předem ručně kompilovat.
```

Comment:
Trochu mi tady chybí informace, k čemu to teda je dobré.

Reply:
Pokud používáš jen několik málo regexů, zůstanou v automatické interní cache, takže to nemusíš dělat ručně. Logicky tedy, pokud jich budeš používat hodně a hodněkrát, může být lepší si je kompilovat a uchovávat ve vlastní režii. Přijde mi to, že bych doplněním textu jen otočil význam poslední věty a napsal v podstatě to samé. Velmi podobná formulace je i v dokumentaci k funkci `compile`.

Reply:
Je otázka, zda je to materiál na seznamovací lekci. Zní to jako velice úzký use case.

OTOH je ale dobré umět to číst :/

Reply:
Důvod, proč je to tu zmíněno je právě ten, že se to často objevuje v tutoriálech, aniž by tam někdo napsal, že to vlastně není potřeba dělat dokud nenarazíš na problémy z výkonem.